lib/xnvme_be_macos_driverkit_dev.c

```diff
@@ -0,0 +1,230 @@
+		XNVME_DEBUG("FAILED: IOServiceGetMatchingServices(); ret(0x%08x), '%s'", ret,
+			    mach_error_string(ret));
+
+		return -ENODEV;
+	}
+
+	tmp_opts.be = xnvme_be_macos_driverkit.attr.name;
+	tmp_opts.nsid = 1;
+
+	while (true) {
+		dev = NULL;
+		memset(&uri, 0, XNVME_IDENT_URI_LEN);
+		service = IOIteratorNext(iterator);
+
+		if (!service) {
+			break;
+		}
+
+		ret = IORegistryEntryGetName(service, uri);
+		if (strncmp("MacVFN-", uri, 7)) {
+			continue;
+		}
+		XNVME_DEBUG("INFO: MacVFN uri matched: %s", uri);
+
+		dev = xnvme_dev_open(uri, &tmp_opts);
+		if (!dev) {
+			XNVME_DEBUG("INFO: xnvme_dev_open(): %d", errno);
+			return -errno;
+		}
+		if (cb_func(dev, cb_args)) {
+			xnvme_dev_close(dev);
+		}
+	}
+
+	IOObjectRelease(iterator);
+
+	return 0;
+}
+
+void
+xnvme_be_macos_driverkit_dev_close(struct xnvme_dev *dev)
+{
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)dev->be.state;
+
+	if (!state->device_service) {
+		XNVME_DEBUG("FAILED: !state->device_service");
+		return;
+	}
+
+	IOServiceClose(state->device_service);
+	IOServiceClose(state->device_connection);
+}
+
+int
+_xnvme_be_driverkit_create_ioqpair(struct xnvme_be_macos_driverkit_state *state, int qd, int flags)
+{
+	unsigned int qid = __builtin_ffsll(state->qidmap);
+	NvmeQueue ioqpair_return;
+	NvmeQueue ioqpair_create;
+	kern_return_t ret;
+	size_t output_cnt;
+
+	if (qid == 0) {
+		XNVME_DEBUG("FAILED: no free queue identifiers");
+		return -EBUSY;
+	}
+
+	qid--;
+
+	XNVME_DEBUG("INFO: nvme_create_ioqpair(%d, %d, %d)", qid, qd + 1, flags);
+
+	// nvme queue capacity must be one larger than the requested capacity
+	// since only n-1 slots in an NVMe queue may be used
+	ioqpair_create = (NvmeQueue){
+		.id = (uint64_t)qid,
+		.depth = (uint64_t)qd + 1,
+		.vector = -1,
+		.flags = (uint64_t)flags,
+	};
+	output_cnt = sizeof(NvmeQueue);
+
+	ret = IOConnectCallStructMethod(state->device_connection, NVME_CREATE_QUEUE_PAIR,
+					&ioqpair_create, sizeof(NvmeQueue), &ioqpair_return,
+					&output_cnt);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectCallStructMethod(NVME_CREATE_QUEUE_PAIR); "
+			    "ret(0x%08x), '%s'",
+			    ret, mach_error_string(ret));
+		return -EAGAIN;
+	}
+	XNVME_DEBUG("INFO: IOConnectCallStructMethod(NVME_CREATE_QUEUE_PAIR); success");
+
+	state->qid_sync = ioqpair_return.id;
+	state->qidmap &= ~(1ULL << qid);
+
+	return qid;
+}
+
+int
+_xnvme_be_driverkit_delete_ioqpair(struct xnvme_be_macos_driverkit_state *state, int qid)
+{
+	size_t output_cnt = 0;
+	NvmeQueue ioqpair_delete = {
+		.id = (uint64_t)qid,
+	};
+	kern_return_t ret;
+
+	ret = IOConnectCallStructMethod(state->device_connection, NVME_DELETE_QUEUE_PAIR,
+					&ioqpair_delete, sizeof(NvmeQueue), NULL, &output_cnt);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectCallStructMethod(NVME_DELETE_QUEUE_PAIR); "
+			    "ret(0x%08x), '%s'",
+			    ret, mach_error_string(ret));
```

Comment:
Is this missing a ``return`` with error?

Reply:
Added return.